### DISCUSSION TITLE
ECWolf mods not working

### DISCUSSION THREAD

**Discussion by leomontenegro6** [wummle/arkos#7] [Q&A]
I'm trying to run some Wolfenstein 3D mods with standalone ECWolf by following [these instructions](https://github.com/christianhaitian/arkos/wiki/ArkOS-Emulators-and-Ports-information#wolfenstein), but I'm getting some issues.

When using the ".ecwolf" extension and running "Scan_for_new_games", no menu entries are created, even after restarting EmulationStation. It's as if the files weren't being detected.

I also tried renaming the extensions from ".ecwolf" to simply ".wolf". In this case, running "Scan_for_new_games" indeed made effect, with the menu entries being created after restarting EmulationStation. However, opening these new menu entries isn't making the mods load up. Instead, the user is redirected back to EmulationStation.

It's worth mentioning that the original games (Wolf3D and SOD) are working just fine with standalone ECWolf. Only mods are failing. Here's the syntax of one of my .wolf files:
```
SUBDIR=Wolfenstein 3D
DATA=WL6
PK3_1=/roms/wolf/wolf3d_coe_v2.pk3
```
Am I doing something wrong on my end? Does anyone had luck with ECWolf mods besides me?

**Comment by leomontenegro6** [comment]
Upon an investigation I did yesterday, according to ArkOS final (last update for RG351P devices, sent by Christian Haitian), this functionality isn't implemented as described by the wiki.

Firstly, the .ecwolf extension isn't accepted, only .wolf is instead.

Secondly, the `ecwolf.sh` file isn't considering the contents of the .wolf files. Instead, it's only getting its filename and using it to search for folders containing WL1/WL6/SOD/N3D files.

I did the following modifications to the `ecwolf.sh`, whose purpose is to check if .wolf has any content (instead of being empty). If true, then it extracts the key/values and uses them to load ECWolf with the corresponding data / pk3 files:

```sh
#!/bin/bash
if [[ $2 == "/roms/wolf/Scan_for_new_games.wolf" ]]; then
    printf "\033c" >> /dev/tty1
    cd /roms/wolf
    ./Scan_for_new_games.wolf
    printf "\n\nFinished scanning the wolf folder for games." >> /dev/tty1
    printf "\nPlease restart emulationstaton to find the new shortcuts" >> /dev/tty1
    printf "\ncreated if any.\n" >> /dev/tty1
    sleep 5
    printf "\033c" >> /dev/tty1
elif [[ $1 == "standalone" ]]; then
    if [ ".$(echo "$2"| cut -d. -f2)" == ".sh" ] || [ ".$(echo "$2"| cut -d. -f2)" == ".SH" ]; then
        # Shell-script file, so simply run it.
        "$2"
    elif [ -s "$2" ]; then
        # File has contents, so try parsing it
        SUBDIR=""; DATA=""; PK3_1=""; PK3_2=""
        sudo dos2unix "${2}"
        while IFS== read -r key value; do
            if [ "$key" == "SUBDIR" ]; then SUBDIR+="${value}"
            elif [ "$key" == "DATA" ]; then DATA+="${value}"
            elif [ "$key" == "PK3_1" ]; then PK3_1+="${value}"
            elif [ "$key" == "PK3_2" ]; then PK3_2+="${value}"
            fi
        done < "${2}"

        cd "/roms/wolf/${SUBDIR%.*}"
        SDL_GAMECONTROLLERCONFIG="03000000091200000031000011010000,OpenSimHardware OSH PB Controller,a:b0,b:b1,x:b3,y:b2,leftshoulder:b4,rightshoulder:b5,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,leftx:a0~,lefty:a1~,leftstick:b8,lefttrigger:b10,rightstick:b9,back:b7,start:b6,rightx:a2,righty:a3,righttrigger:b11,platform:Linux," /opt/ecwolf/ecwolf --savedir "/roms/wolf/${SUBDIR%.*}" --file "$PK3_1" "$PK3_2"
    else
        # File is empty, so simply get the filename and assume it points to a folder that has a wolf3d game.
        filename="$(basename "$2")"
        cd "/roms/wolf/${filename%.*}"
        SDL_GAMECONTROLLERCONFIG="03000000091200000031000011010000,OpenSimHardware OSH PB Controller,a:b0,b:b1,x:b3,y:b2,leftshoulder:b4,rightshoulder:b5,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,leftx:a0~,lefty:a1~,leftstick:b8,lefttrigger:b10,rightstick:b9,back:b7,start:b6,rightx:a2,righty:a3,righttrigger:b11,platform:Linux," /opt/ecwolf/ecwolf --savedir "/roms/wolf/${filename%.*}"
    fi
else
    filename="$(basename "$2")"
    game=$(find "/roms/wolf/${filename%.*}" -type f -iname "*.exe" | grep -iv "install" | grep -iv "setup")
    /usr/local/bin/retroarch -L /home/ark/.config/retroarch/cores/ecwolf_libretro.so "$game"
fi
```

This is WIP. It worked for me, with a few caveats:

1. The .wolf must be in Unix (LF) text encoding (although there's a step that tries converting the encoding, through the `dos2unix` command, but it wasn't present on ArkOS final by default, so I had to manually install it with `sudo apt install dos2unix`).
2. There should exist a line break below the last line, otherwise, the last parameter will be possibly skipped.

I'm not sure about how exactly it was implemented in other CFW such as AmberElec or JELOS. Maybe we won't need the DATA parameter, because this can be obtained from the SUBDIR parameter (since the script goes to that folder and runs ecwolf inside it).

I'm willing to contribute these changes to wummle/arkos. Any thoughts or suggestions about that implementation?

**Reply by wummle** [reply]
Looks good

**Comment by Slayer366** [comment]
My apologies for not having seen this sooner.  I was unaware that dos2unix was not already installed in ArkOS on the 351P/351M.  As of 2024-09-08, dos2unix will now get installed via the OTA update!
